File: Haskell/Calendario.hs
{-------------------------
    Programa: Calendario.hs
    Funcion: Imprime el calendario de x anio

    Nombre: Arellano Ochoa Daniel Ignacio
    No. de control: 19141118
-------------------------}
import Text.Printf
main = do
    printf "Bienvenido, que calendario desea? \n"
    lectura <- getLine
    let 
        a = read lectura :: Int
        d = diaInicial a
    impMes 1 d a

diaInicial :: Int -> Int
diaInicial a = primerDia a 2022 6

primerDia :: Int -> Int -> Int -> Int
primerDia a aActual dActual 
    | (a == aActual) = dActual
    | ((a > aActual) && ((esBisiesto aActual) == True)) = primerDia a (aActual + 1) (mod (dActual + 2) 7)
    | ((a > aActual) && ((esBisiesto aActual) == False)) = primerDia a (aActual + 1) (mod (dActual + 1) 7)
    | ((a < aActual) && ((esBisiesto (aActual - 1)) == True)) = primerDia a (aActual - 1) (mod (dActual - 2) 7)
    | ((a < aActual) && ((esBisiesto (aActual - 1)) == False)) = primerDia a (aActual - 1) (mod (dActual - 1) 7)
    | otherwise = 0
  
esBisiesto :: Int -> Bool
esBisiesto a 
    | ((mod a 4 == 0) && (mod a 100 == 0) && (mod a 400 == 0)) = True
    | ((mod a 4 == 0) && (mod a 100 /= 0)) = True
    | otherwise = False

impMes :: Int -> Int -> Int -> IO()
impMes mes d a 
    | (mes == 12) = do 
        impEncabezado a mes d
        impSemana d 1 31
    | (mes == 1 || mes == 3 || mes == 5 || mes == 7 || mes == 8 || mes == 10) = do
        impEncabezado a mes d
        impSemana d 1 31
        impMes (mes + 1) (mod (d + 31) 7) a
    | (mes == 4 || mes == 6 || mes == 9 || mes == 11) = do
        impEncabezado a mes d
        impSemana d 1 30
        impMes (mes + 1) (mod (d + 30) 7) a
    | otherwise = do
        impEncabezado a mes d
        if(esBisiesto a)
            then do
                impSemana d 1 29
                impMes (mes + 1) (mod (d + 29) 7) a
            else do
                impSemana d 1 28
                impMes (mes + 1) (mod (d + 28) 7) a
    
impEncabezado :: Int -> Int ->  Int -> IO()
impEncabezado a mes dInicio = do
    let mesLetra = nombreMes mes
    printf "%s de %d \n D  L Ma Mi  J  V  S \n" mesLetra a
    impEspacios dInicio

impEspacios :: Int -> IO()
impEspacios dInicio
    | dInicio > 0 = do
        printf "   "
        impEspacios (dInicio - 1)
    | otherwise = printf ""

impSemana :: Int -> Int -> Int -> IO()
impSemana dActual fechaCon lim 
    | (lim > 0 && dActual == 6) = do
        impDia fechaCon 
        printf "\n"
        impSemana 0 (fechaCon + 1) (lim - 1)
    | (lim > 0 && dActual /= 6) = do 
        impDia fechaCon 
        printf " "
        impSemana (dActual + 1) (fechaCon + 1) (lim - 1)
    | otherwise = printf "\n \n"

impDia :: Int -> IO()
impDia fechaCon
    | fechaCon > 9 = printf "%d" fechaCon
    | otherwise = printf " %d" fechaCon

nombreMes :: Int -> String
nombreMes mes = nombreMes2 mes ["Enero", "Febrero", "Marzo", "Abril", "Mayo", "Junio", "Julio", "Agosto", "Septiembre", "Octubre", "Noviembre", "Diciembre"]

nombreMes2 :: Int -> [String] -> String
nombreMes2 mes (x : resto) 
    | mes == 1 = x
    | otherwise = nombreMes2 (mes - 1) resto

File: Prolog/Parcial 1/Calendario.pl

main :-
    write('Bienvenido, que calendario desea? '), read(A), nl,
    diaInicial(A,D),
    impM(1,D,A).

impM(Mes,D,A):-
    impTitulo(Mes, D),
    (Mes=12)->
        imprimirSemanas(D,1,31),
        true
    ;(Mes=1 ; Mes=3 ; Mes=5; Mes=7 ; Mes=8 ; Mes=10)->
        imprimirSemanas(D,1,31),
        Mes2 is Mes+1,
        D2 is (D+31) mod 7,
        impM(Mes2,D2,A)
    ;(Mes=4 ; Mes=6 ; Mes=9 ; Mes=11)->
        imprimirSemanas(D,1,30),
        Mes2 is Mes+1,
        D2 is (D+30) mod 7,
        impM(Mes2,D2,A)
    ;(esBisiesto(A))->
        imprimirSemanas(D,1,29),
        Mes2 is Mes+1,
        D2 is (D+29) mod 7,
        impM(Mes2,D2,A)
    ;
        imprimirSemanas(D,1,28),
        Mes2 is Mes+1,
        D2 is (D+28) mod 7,
        impM(Mes2,D2,A).


impTitulo(Mes, Dinicio):-
    nombreMes2(Mes,X),
    write(X),nl,
    write(' D  L Ma Mi  J  V  S'), nl,
    imprimirEspacios(Dinicio).

imprimirEspacios(Dinicio) :-
    (Dinicio \= 0) -> 
        write('   '),
        Dinicio2 is Dinicio-1,
        imprimirEspacios(Dinicio2)
    ;true.

imprimirSemanas(Dact, Fcon, Lim):-
    (Lim > 0, Dact = 6) ->
        imprimirDia(Fcon), nl,
        Lim2 is Lim-1,
        Fcon2 is Fcon+1,
        imprimirSemanas(0,Fcon2,Lim2)
    ;(Lim > 0, Dact \= 6)->
        imprimirDia(Fcon),write(' '),
        Lim2 is Lim-1,
        Fcon2 is Fcon+1,
        Dact2 is Dact+1,
        imprimirSemanas(Dact2,Fcon2,Lim2)
    ;nl,nl,true.

imprimirDia(Fcon):-
    (Fcon>9)->
        write(Fcon)
    ;
        write(' '),write(Fcon),true.

diaInicial(A,D) :-
    primerDia(A,D,2022,6).

primerDia(A,D,A,D) :- !.
primerDia(A,D,Aact,Dact) :-
    A > Aact, !,
    Aact1 is Aact+1, ajusteBis(Aact,X),
    Dact1 is (Dact+1+X) mod 7,
    primerDia(A,D,Aact1, Dact1).
primerDia(A,D,Aact,Dact) :-
    A < Aact, !,
    Aact1 is Aact-1, ajusteBis(Aact1,X),
    Dact1 is (Dact-1-X) mod 7,
    primerDia(A,D,Aact1, Dact1).

ajusteBis(A,1) :- 
    esBisiesto(A), !.
ajusteBis(_,0).

esBisiesto(A) :-
    R is A mod 400, R=0, !.
esBisiesto(A) :-
    R is A mod 4, R = 0, !,
    R1 is A mod 100, R1\=0.

nombreMes2(M,X) :-
    nombreMes3(M, ['Enero','Febrero','Marzo','Abril','Mayo','Junio','Julio','Agosto','Septiembre','Octubre','Noviembre','Diciembre'], X).

nombreMes3(1,[X | _],X) :- !.
nombreMes3(M,[_ | Resto],X) :- 
    M1 is M-1, nombreMes3(M1, Resto, X).